content/blog/eng/how-i-became-a-gitlab-contributor-to-fix-an-issue-in-our-projects-cicd-pipeline.md

```diff
@@ -0,0 +1,322 @@
+
+Having to deal with a large amount of different protocols/formats makes everything error-prone: what if a certain company decides that a property is now a `number` instead of a `string` but they fail to let us know that their specifications changed?
+
+We can't afford throwing errors to the end user because of such a small change, so we need a better way to be somehow proactive and intercept those changes before they reach our production systems.
+
+For this reason we have a set of multiple tests that run in our **CI/CD** environment and that help us keeping everything in check.
+Today we're going to focus on **unit tests** only.
+
+## Unit tests and **Docker Compose**
+
+In **DevOps** world, **unit tests** can be seen as an automated way to do **regression testing**, meaning that we can use them to guarantee that our new commit didn't break an already working feature.
+
+This is extremely useful for projects that start to outgrow their team and that would require a significant effort to manually verify everything before committing.
+
+In our scenario this means that we defined these **unit tests** in order to validate the request/response pair's format for each insurance company, thus being able to quickly react to their changes.
+
+Since **unit tests** are not meant to depend on external systems, however, we decided to build another layer: **mock servers**.
+
+Here's a quick chart showing a simplified version of our testing process including just two companies (I promise it's not that hard, it's just Mermaid some issues with circular layouts 🙂):
+
+[![](https://mermaid.ink/img/pako:eNqVVFFvmzAY_CuWnwsd3tZuedhUkaqK0pZIqyZNsAcPvjRWAVPbtIpC_vs-Y6CUJp3Gk7k7H9zHmR1NZQZ0Ru8Vrzbkbp6UBK-L1SoWpfbuuYFnvv3t0PnlzziDJ8hlBeoFY2_Aq8VdfCXMNf_TAeEiDhen4by7vYnCZRAXMn3wUllUvNx6wYhirynWUdEyNqANsXgOBrIOX0aBI9Zc5JCRqeUyYgf53jeMblYXt7_iDie6glSsRcqNkKVG0ZCUeN63JvBJVetNY1MOcVuG-UTVZYNp-9Qt_LGDbbQpYd-rseMe5k48H6lPuAce6z5uG8c5BP9WMqdko2E77Wer1RWmgrHt8Ph213Epm74p6s588sRzkWFRDnp3pFTBTmh3933_HxZsZMHeWLy4t05b0A3W5CBXysY2ZcKxd_ax0T7WtwAtWvjczqeSCr-ePQJD046TdtWyXyxrlMBD0_Tlm0i--qSu7EBetcZJjwvsylX0w6ij9IQWoAouMjzmO6tMqNlAAQmd4TLj6iGhSblHnbO8zASmp7M1zzWcUF4b-WNbpnRmVA29aC44_jKKTrX_C-c4VcU)](https://mermaid.live/edit#pako:eNqVVFFvmzAY_CuWnwsd3tZuedhUkaqK0pZIqyZNsAcPvjRWAVPbtIpC_vs-Y6CUJp3Gk7k7H9zHmR1NZQZ0Ru8Vrzbkbp6UBK-L1SoWpfbuuYFnvv3t0PnlzziDJ8hlBeoFY2_Aq8VdfCXMNf_TAeEiDhen4by7vYnCZRAXMn3wUllUvNx6wYhirynWUdEyNqANsXgOBrIOX0aBI9Zc5JCRqeUyYgf53jeMblYXt7_iDie6glSsRcqNkKVG0ZCUeN63JvBJVetNY1MOcVuG-UTVZYNp-9Qt_LGDbbQpYd-rseMe5k48H6lPuAce6z5uG8c5BP9WMqdko2E77Wer1RWmgrHt8Ph213Epm74p6s588sRzkWFRDnp3pFTBTmh3933_HxZsZMHeWLy4t05b0A3W5CBXysY2ZcKxd_ax0T7WtwAtWvjczqeSCr-ePQJD046TdtWyXyxrlMBD0_Tlm0i--qSu7EBetcZJjwvsylX0w6ij9IQWoAouMjzmO6tMqNlAAQmd4TLj6iGhSblHnbO8zASmp7M1zzWcUF4b-WNbpnRmVA29aC44_jKKTrX_C-c4VcU)
```

Comment:
We should include images statically, so it wouldn't break in the future

Reply:
Yeah, it's just a placeholder in case I want to modify the chart again before publishing everything. I'll replace it with the definitive version later.